### DISCUSSION TITLE
Is anyone working on a python query builder or ORM of some kind?

### DISCUSSION THREAD

**Discussion by autoferrit** [geldata/gel#1713] [General]
I am curious if anyone knows of any work on making a query builder for edgedb. If not, perhaps we could fork an existing builder to start from?

**Comment by elprans** [comment]
@agritheory has made a proof-of-concept mapper some time ago: https://gitlab.com/agritheory/edgewise

**Reply by agritheory** [reply]
The 'testing' branch is actually the more mature of the two and I should have merged the changes into master. @autoferrit I could use a hand with refactoring away from string interpolation in the queries if that's interesting to you.

**Reply by autoferrit** [reply]
Hmm. interesting. possibly. I've never worked on anything like this, but im interested. what should I look at?

**Reply by autoferrit** [reply]
ah you're using attrs. I was actually considering basing off of pydantic to be able to infer types. But I guess for me that makes sense as I am also trying to use fastapi.

**Reply by agritheory** [reply]
> Hmm. interesting. possibly. I've never worked on anything like this, but im interested. what should I look at?

The place that needs a refactor is here: https://github.com/agritheory/edgewise/blob/master/edgewise/queries.py 

In particular I'm hoping that refactoring to variables would solve for string values that contain reserved words and special characters.

The other problem that needs attention is handling links: there needs to be some sort of `SELECT id IF EXISTS ELSE INSERT` logic but I haven't figured it out.

**Reply by agritheory** [reply]
> ah you're using attrs. I was actually considering basing off of pydantic to be able to infer types. But I guess for me that makes sense as I am also trying to use fastapi.

I have no experience with pydantic but I am able to load the type from EdgeDB's types into the attrs class. I'm not sure, but I think this should yield a similar result.

**Reply by elprans** [reply]
>  there needs to be some sort of `SELECT id IF EXISTS ELSE INSERT` logic but I haven't figured it out.

The latest release supports this directly via the new `INSERT ... UNLESS CONFLICT ELSE SELECT` construct.

**Reply by autoferrit** [reply]
Ok. yea, maybe at some point I can try and see if using pydantic works as well. possibly an optional requirement, like when installing with `pip install edgewise[pydantic]` or something.
Another thing I was thinking of, is the api of using the library. I really like how the new querybuilder is working in the upcoming `SQLAlchemy` version where you build your query like 
```python
results = select(User).filter(id=5).first()
```
I had made an example of how I would see that, based on the examples on the edgedb website. It's messy but this is what I had (im sure there is cleaner ways to do this). This is also glossing over a lot of things, as it was more of a thought experiment.
```python
class User:
    pass

class Document:
    def __call__(self, cls):
        return self

    def select(self, keys=None):
        return self

    def filter(self, *args, **kwargs):
        return self

class Owner:
    name = None

    def __init__(self, name):
        self.name = name

class Issue:
    number = 5

    def time_spent_log(self):
        pass

    time_spent_log.__setattr__("spent_time", True)

num = 5
document = Document()
doc = (
    document(User)
    .select(
        [
            "number",
            "title",
            {"owner": Owner.name},
            {"watchers": ["name"]},
            {"priority": ["name"]},
            {"status": ["name"]},
            {
                "total_time_spent": document("sum").select(
                    Issue.time_spent_log.spent_time
                )
            },
            "start_date",
            "due_date",
            {"related_to": ["number", "title", {"priority": ["name"]}]},
        ]
    )
    .filter(number=num)
)
```

**Reply by autoferrit** [reply]
I might be willing to help with it though. I definitely want to try and use edgedb anyways.
